src/WCSimPrimaryGeneratorAction.cc

```diff
@@ -1552,6 +1558,63 @@ void WCSimPrimaryGeneratorAction::GeneratePrimaries(G4Event* anEvent)
       SetBeamDir(dir);
       SetBeamPDG(pdg);
 
+    } else if (useHepMC3Evt) {
+
+      G4cout << "Using HepMC3 event" << G4endl;
+      // Check if the WCSimNuHepMC3Reader object has been initiaited yet
+      if (!hepmc3_reader) {
+        hepmc3_reader = new WCSimNuHepMC3Reader(hepmc3_filename, myDetector);
+      }
+
+      // Get the next event from the reader
+      hepmc3_reader->ReadEvent(hepmc3_positionGen);
```

Comment:
`ReadEvent()` returns a `bool` to say whether it read the event successfully or not (technically whether there are any particles or not, but I presume it's the same thing)

I wonder if you can add a check here, and if there is no event, to call
```
 G4RunManager::GetRunManager()-> AbortRun();
```
(as is done for muline when the file is exhausted)

I'm thinking of the situation where the input file has 10 events, but the WCSim user has requested 1000 events via `beamOn`. This will be slow...